packages/compiler-cli/src/ngtsc/testing/src/utils.ts

```diff
@@ -138,3 +138,12 @@ function bindingNameEquals(node: ts.BindingName, name: string): boolean {
   }
   return false;
 }
+
+export function getSourceCodeForDiagnostic(diag: ts.Diagnostic): string {
+  if (diag.file === undefined || diag.start === undefined || diag.length === undefined) {
+    throw new Error(
```

Comment:
I believe this is an internal message (i.e. framework users will not see this), but I think we can still improve it a bit. Currently it's unclear what file/start/length are.
```suggestion
        throw new Error(
          'Unable to get source code for diagnostic. Provided diagnostic instance doesn't contain "file", "start" and/or "length" properties.');
```

Reply:
We actually tried earlier, but printing the diagnostic or file would give `[object Object]` and `JSON.stringify()` fails because it s a circular structure. Probably don't need to worry too much for an assertion error like this and don't need to spend too much effort trying to make it nice.

Reply:
Yeah, using `JSON.stringify()` might not be possible. My proposal is to rephrase the error string to be more explicit on what the problem is (currently it's quite unclear) without additional `JSON.stringify()` calls.